### PR TITLE
Collective selection: select and call phases separation

### PR DESCRIPTION
**Pull request by AlexeyMalkhanov**
Collective selection logic. This PR includes separation of selection and call phases for Barrier, Bcast, Reduce and Allreduce collectives. This infrastructure doesn't depend on collective framework.

### CODE REVIEW

File: src/mpid/ch4/netmod/ofi/ofi_coll.h [hunk 1]
```diff
@@ -12,20 +12,52 @@
...
+#include "ch4_coll_params.h"
+#include "ofi_coll_select.h"
 
 #undef FUNCNAME
 #define FUNCNAME MPIDI_NM_mpi_barrier
 #undef FCNAME
 #define FCNAME MPL_QUOTE(FUNCNAME)
-static inline int MPIDI_NM_mpi_barrier(MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag)
+static inline int MPIDI_NM_mpi_barrier(MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag,
+                                       void * ch4_algo_parameters_container_in)
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_coll.h]: The algo params argument is a `void *` here but `MPIDI_OFI_coll_algo_container_t` in the POSIX function headers. Please make them consistent, preferably with the correct type to avoid casting immediately.

File: src/mpi/coll/bcast.c [hunk 2]
```diff
@@ -121,16 +121,16 @@ int MPI_Bcast(void *buffer, int count, MPI_Datatype datatype, int root, MPI_Comm
...
-static int MPIR_Bcast_binomial(
+int MPIR_Bcast_binomial(
     void *buffer, 
     int count, 
     MPI_Datatype datatype, 
     int root, 
     MPIR_Comm *comm_ptr,
     MPIR_Errflag_t *errflag)
 {
-    int        rank, comm_size, src, dst;
```
Comment (raffenet) [inline src/mpi/coll/bcast.c]: unrelated whitespace change

File: src/mpi/coll/bcast.c [hunk 3]
```diff
@@ -1567,7 +1567,7 @@ int MPI_Bcast( void *buffer, int count, MPI_Datatype datatype, int root,
     /* ... body of routine ...  */
     
     mpi_errno = MPID_Bcast( buffer, count, datatype, root, comm_ptr, &errflag );
-    if (mpi_errno) goto fn_fail;
+    if (mpi_errno) MPIR_ERR_POP(mpi_errno);
```
Comment (raffenet) [inline src/mpi/coll/bcast.c]: what's the reason for changing these `goto`s? I'm not saying they are wrong, just want to understand better.

File: src/mpi/coll/bcast.c (hunk 2 above)
Comment (AlexeyMalkhanov) [inline src/mpi/coll/bcast.c]: Will correct that

File: src/mpi/coll/bcast.c (hunk 3 above)
Comment (AlexeyMalkhanov) [inline src/mpi/coll/bcast.c]: MPIR_ERR_POP is a wrapper for "goto fn_fail" with additional capabilities to handle errors. In case when HAVE_ERROR_CHECKING is defined MPIR_ERR_POP calls MPIR_Err_create_code(...) and then "goto fn_fail". If HAVE_ERROR_CHECKING is not defined then MPIR_ERR_POP is just "goto fn_fail".

File: src/mpid/ch4/netmod/ofi/ofi_coll.h (hunk 1 above)
Comment (AlexeyMalkhanov) [inline src/mpid/ch4/netmod/ofi/ofi_coll.h]: MPIDI_NM_mpi_barrier is the general netmod function which have its own implementations for each particular netmod: OFI, UCX, etc. To hide the details about particular netmod in the general NM functions we pass void * as a parameter. For netmod specific functions like MPIDI_OFI_Barrier_recursive_doubling we pass  MPIDI_OFI_coll_algo_container_t * since it is clear from the prefix "MPIDI_OFI" of the name of the function which type of te container shoul be used. The same idea was used for SHM/POSIX: in shm_coll.h we pass void * as a parameter for MPIDI_SHM_mpi_barrier function (since it is general), but in posix_coll.h we pass MPIDI_POSIX_coll_algo_container_t * for MPIDI_POSIX_mpi_barrier. 
The difference between NM and SHM is that NM doesn't have separate nm_coll.h with general for all netmods functions while SHM has.

File: src/mpi/coll/bcast.c (hunk 3 above)
Comment (raffenet) [inline src/mpi/coll/bcast.c]: :+1:

**Comment by raffenet** [comment]
@AlexeyMalkhanov Could you separate out the follow parts into their own commits from this PR?
1. new collective algorithms
2. NM interface changes
3. ch4/ofi coll selection changes

**Comment by raffenet** [comment]
In the meantime I'm going to test this version of the branch to smoke out any obvious problems.
test:jenkins/ch4/most

**Comment by AlexeyMalkhanov** [comment]
@raffenet I'm working at the separation. Will make new PR asap

**Comment by wesbland** [comment]
Don't make a new PR @AlexeyMalkhanov, just force push to this branch.

**Comment by AlexeyMalkhanov** [comment]
Working at conflicts resolving

**Comment by AlexeyMalkhanov** [comment]
@raffenet could you please launch tests on the commit with the conflict. I have resolved it but would not push it until I verify all build configurations. Will update the PR asap.

**Comment by raffenet** [comment]
test:jenkins/ch4/most

**Comment by AlexeyMalkhanov** [comment]
@raffenet I've resolved the conflict and pushed the updated PR

**Comment by ddurnov** [comment]
test:jenkins/ch4/most

**Comment by AlexeyMalkhanov** [comment]
@raffenet Could you please launch tests on the updated version?

**Comment by raffenet** [comment]
test:jenkins/ch4/most

**Comment by raffenet** [comment]
It looks like this version segfaults when calling any of the reduce algos in ch4/ucx.

**Comment by raffenet** [comment]
@AlexeyMalkhanov can you also `dos2unix` all the `*_coll_globals_default.c` files?

**Comment by AlexeyMalkhanov** [comment]
@raffenet I've removed the wrong line endings in the mentioned files and updated the PR. As far as issues with reduce algos in ch4/ucx are concerned, I could not reproduce them. Could you please provide me with the details?

**Comment by raffenet** [comment]
I will try to reproduce these now. If I can't I will ignore them and move on with the merge.

**Comment by raffenet** [comment]
Hmm, nevermind. A closer look at the Jenkins results doesn't show the reduce problems. Must have been from another job.

**Comment by raffenet** [comment]
Superseded by #2726. Closing.
